### ISSUE TITLE
Mosh usually/always fails to get MOSH CONNECT when connecting to a server running Dropbear

### ISSUE THREAD

**Issue by cgull** [mobile-shell/mosh#819]
When mosh is used to connect to a server running `dropbear` instead of OpenSSH's `sshd`, the client will rarely or never receive the `MOSH CONNECT` message.  The ssh connection closes before any output is received from mosh-server.  It appears that Dropbear doesn't tolerate the `-n -tt` flags very well (this is a somewhat nonsensical combination-- "don't open stdin" and "use a pty for interactive usage").  sshd seems to handle this fine.

@cjdelisle discovered this on a MIPS Creator C120 SBC running something called "Debian 8" with ab apt source of `deb http://ftp.uk.debian.org/debian/ jessie-updates main`, `mosh/stable,now 1.2.4a-1+b2 mipsel` installed, and a kernel bugfixed for a VM issue breaking mmap(), uname is `Linux ci20-3006 3.18.3+ #4 SMP PREEMPT Sat Mar 7 10:25:33 EST 2015 mips GNU/Linux`.  But I've also reproduced this on an Ubuntu 16.10 x86_64 install.

Workarounds/fixes include deleting `-n` from the ssh command line generated by the `mosh` script, or deleting `-tt` from the same command line (only works with `mosh-server` 1.2.5+), or using a very current version of Mosh which has the `--no-ssh-pty` option.

Dropbear could do better in this area, but I'm not sure it's an actual bug, because of the oddness of the combination of options, and the way the ssh2 protocol does independent reliable unidirectional byte streams for stdin/stdout/stderr which can be opened/closed independently (actually, when `-n` is used, it doesn't create one for stdin at all, if I remember right).

We might be able to just drop the `-n` flag, I think this might be a remnant from when the script used a pty.

**Comment by cgull** [comment]
A few more things:
Simple repro of Dropbear issue:
```shell
sudo dropbear -p 2222
ssh -n -tt -p 2222 localhost date
```
The C120 has `dropbear/stable,now 2014.65-1`
The Ubuntu instance has `dropbear/yakkety,yakkety,now 2016.74-1` and `mosh/yakkety,now 1.2.6-1build1`

**Comment by cgull** [comment]
A few more things:
Simple repro of Dropbear issue:
```shell
sudo dropbear -p 2222
ssh -n -tt -p 2222 localhost date
```
The C120 has `dropbear/stable,now 2014.65-1`
The Ubuntu instance has `dropbear/yakkety,yakkety,now 2016.74-1` and `mosh/yakkety,now 1.2.6-1build1`

**Comment by andersk** [comment]
See also #464, #524.

**Comment by cgull** [comment]
Editing the Mosh Perl script to remove the '-n' option mostly resolved the original problem.  But both @cjdelisle and I saw that mosh-1.2.6 `mosh-server` was still not 100% reliable, not even with Mosh from Git as client and `--no-ssh-pty`.  In testing, even with `ssh -T -p 2222 localhost mosh-server new` *or* `ssh -tt -p 2222 localhost mosh-server new`, text was still being randomly dropped occasionally, and sometimes `MOSH CONNECT` was dropped.  I found that the fix for #809 also resolves these problems on servers running Dropbear.  Mosh from Git `mosh-server` appears to be reliable for both these commands.

Bottom line:

If you're running Dropbear on your server and you need reliable connections, you must:
* switch to OpenSSH `sshd`, or
* use Mosh from Git after commit  70d48fbcc038d4f746bab029e8449812263f2cf9 (or mosh-1.2.7, when that comes out fairly soon), on *both* client and server, and use the `--no-ssh-pty` option on your client.
If you only upgrade your client and use `--no-ssh-pty`, it will work most of the time, but not always.

I don't think there's anything more Mosh can to do fix this, but I'm open to any ideas :)
Many thanks to @cjdelisle for providing access to his system!

**Comment by cgull** [comment]
Edited previous comment, there is one fix to try.

**Comment by cgull** [comment]
@glance-: What was your motivation for adding `-n` in the first place (e0dfe363a0fefab15aa461c4fce00af3b35d8adb)?  Do you think we still need it?

**Comment by glance-** [comment]
It was to make sure ssh didn't eat up commands  / keypresses ment for the shell started via the mosh session.

IE. Its quite common that i type mosh server and start typing on the command that i'm going to run on that server before the session is estabished. If the `-n` isn't there, the ssh command will "eat" those caracters and they will not stay buffered in the terminal emulatior until the shell is started inside the mosh session.


The current  `-n`-behavior is consistent with how the same scenario with ssh would play out.

**Comment by glance-** [comment]
The `-n -tt` is a very valid combination. Some old mosh-server's segfaulted they didn't have a tty which they could query for size. For the `-n` part, see above.

**Comment by cjdelisle** [comment]
So what you're saying is that dropbear is not supported ?  In this case it probably ought to be documented.

**Comment by glance-** [comment]
No. Read it again. I said nothing about dropbear. But what can be concluded from this conversation is that dropbear isn't compatible with the defacto standard of OpenSSH. The workaround with using `--no-ssh-pty` seams sensible to couple with dropbear's difference here.

I would argue that keeping the behavior that mosh/ssh doesn't "eat" up early input is more important than coulping with dropbears oddities. Those few users can use `--no-ssh-pty` and a modern enough mosh-server.

What would be even better would be of we could autodetect this and try to do the right thing, but i think its hard to detect this from the mosh script, without hurting all users just to make a few dropbear servers happy. Maybe one could first open a tcp connection to the server and just read the banner, and from there make the choice on what to do.

**Comment by cjdelisle** [comment]
Ok, I initially read it to mean "dropbear is broken, not our job to do a workaround" which obviously means not-supported. As far as what is the best way to fix, personally I would be inclined to attempt the connection in `--no-ssh-pty` mode and then if the other side is too old to support that, shutdown and reconnect using "legacy" mode, then at least the issue would be sorted once everybody upgrades, and also it would probably clear up most other lingering ssh compatibility issues. One could even imagine a further fallback which drops the `-n` if it sees that it's connecting to dropbear but I understand that it falls down on you to support all of these fallbacks and probes so I don't want to ask too much.

**Comment by cgull** [comment]
Thanks, @glance-, for describing why it's there.

I reverted my earlier commit to eliminate `-n`.  We've got two somewhat-peripheral issues competing here; I chose to prefer the existing behavior since there are workarounds for Dropbear.

I think this is fixable with a little more effort.  Bringing back `IO::Pty` might help, I'd guess our previous platform issues with that module are resolved by now.

@cjdelisle: Multiple ssh connection attempts have issues like added delay, error handling, UX, and authentication.  I think the best path here is to attempt only one ssh connection.

**Comment by netroby** [comment]
mosh --no-ssh-pty should be as the default configure. or can we provided a default configure file for mosh 

like ~/.moshrc

```
--no-ssh-pty
```

**Comment by cgull** [comment]
@netroby, no, we won't change the default configuration-- as already discussed, it breaks existing functionality for the relatively few users of an unusual ssh server.

So far we've avoided configuration files for Mosh-- it makes Mosh's security story a good deal simpler, and every time the topic has come up, there's been a certain amount of bikeshedding.  We may have to do it at some point, though.

**Comment by cgull** [comment]
User 'opal' on IRC asked about a problem with `tinysshd`, he couldn't establish a session with it, but openssh's sshd works for him.  It's essentially the same thing, though tinysshd's behavior is a little different, I think.  Same workaround, too.

**Comment by travisboss** [comment]
I did as suggested and swapped to openssh and fixed immediately.

**Comment by ssb22** [comment]
I also need `alias mosh='mosh --no-ssh-pty'` to connect to a Raspberry Pi running Raspbian 10's Dropbear package.  It also works when connecting to other servers, so I might as well leave that alias set in my `.bashrc`.

The `--no-ssh-pty` option is present in the version of the `mosh` client installed by Fedora 32, but it is *not* present in the version of the `mosh` client installed by Termux on Android.  Both versions of `mosh` report their version numbers as 1.3.2, but the Fedora version is a Perl script and the Termux version is an ELF binary; I'm not sure what's going on here, but it seems I cannot Mosh from the Android phone to the Raspberry Pi (at least not without installing a less-lightweight SSH server on the Pi) because `--no-ssh-pty` is not currently available in Termux.

**Comment by keithw** [comment]
Termux is shipping an unofficial C++ version of the "mosh" Perl script; you should probably take this up with them.

**Comment by brbsix** [comment]
I had the same issue connecting to my tinysshd servers. As far as I can tell, it was as a result of systemd's socket activation killing off the entire cgroup after disconnecting. I set `KillMode=process` but that alone did not work.

So I wrote a wrapper around `mosh-client` and `mosh-server` to work around this. I now have the script installed at `~/.local/bin/mosh` on Termux and all of my other servers. Works like a charm.

For anyone interested, it's as follows:

    #!/bin/bash

    set -eu

    output=$(ssh "$1" -- echo '$SSH_CONNECTION'\; mosh-server new -l "LANG=$LANG" 2>/dev/null)

    command=$(awk '{if (NR==1) IP=$3; if (NR==2) {MOSH_KEY=$4; PORT=$3}} END{if (MOSH_KEY=="" || IP=="" || PORT=="") exit 1; print "MOSH_KEY="MOSH_KEY, "mosh-client", IP, PORT}' <<<"$output")

    eval "$command"

It assumes that you have your `~/.ssh/config` configured correctly and are able to run `ssh DESTINATION` without any additional options. You run the script by calling `mosh DESTINATION`. I'm now able to connect to my tinysshd servers via Mosh from both Termux and my other Linux machines.

**Comment by xavierog** [comment]
I have just hit this issue while targeting a tinysshd server. I am glad to learn there is a trivial workaround.

FYI and FWIW:
Server-side, tinysshd does not even attempt to fork a mosh-server process.
My strace-based investigations reflected a tinysshd process exiting with status **111** after:
- readlink()'ing `/proc/self/fd/5` as `/dev/pts/0 (deleted)` -- this file descriptor was closed by the parent process a handful of syscalls before the readlink() attempt.
- stat()'ing `/dev/pts/0 (deleted)` and, naturally, getting ENOENT
- listing entries under /dev/pts/ then /dev

This seems to match ttyname() and maybe login_tty().

Client-side, I noticed something funny:

- `mosh targetserver command` fails
- `strace  -o /dev/null mosh targetserver command` fails
- `strace -f -o /dev/null mosh targetserver command` succeeds -- it seems slowing down the SSH client process works around the issue
- `mosh --ssh='ssh -v' targetserver command` sometimes succeeds and sometimes fails.

When the latter does not work, the **111** status code is reflected client-side:
```
Authenticated to 1.2.3.4 (via proxy) using "publickey".
debug1: channel 0: new session [client-session] (inactive timeout: 0)
debug1: Entering interactive session.
debug1: pledge: filesystem
debug1: Sending environment.
debug1: channel 0: setting env LANG = "en_US.UTF-8"
debug1: channel 0: setting env LC_COLLATE = "C"
debug1: Sending command: mosh-server 'new' '-c' '256' '-s' '-l' 'LANG=en_US.UTF-8' '-l' 'LANGUAGE=' '-l' 'LC_COLLATE=C' '--' 'command'
debug1: client_input_channel_req: channel 0 rtype exit-status reply 0
debug1: channel 0: free: client-session, nchannels 1
Connection to 1.2.3.4 closed.
Transferred: sent 3840, received 1952 bytes, in 0.1 seconds
Bytes per second: sent 59112.5, received 30048.9
debug1: Exit status 111
/usr/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```

My hunch is that [tinyssh's channel_forkpty() function](https://github.com/janmojzis/tinyssh/blob/master/tinyssh/channel_forkpty.c#L119) suffers from a race condition between parent and child processes. But @janmojzis 's opinion on this would be more valuable than mine.

**Comment by janmojzis** [comment]
Hi,
fix for the race condition in tinysshd's channel_forkpty() function is here:  https://github.com/janmojzis/tinyssh/compare/master...forkptyfix.

Can You please independently test mosh with this fixed version tinysshd?
Jan

**Comment by xavierog** [comment]
Hi @janmojzis,

First, thanks for your swift reply and patch.

> Can You please independently test mosh with this fixed version tinysshd?

I did.
Good news: the race condition itself seems fixed. The tinysshd child process keeps going: utmp stuff, reading passwd entries, chdir()'ing into the adequate home directory, turning into "/bin/bash" "-c" "mosh-server ..." then "mosh-server".
Bad news: at some point, the tinysshd parent process closes its ptmx file descriptor, which sends a SIGHUP to the child process, which gets killed, resulting in a "-1" exit status being reported client-side:
```console
$ mosh --ssh='ssh -v' targetserver command
[...]
debug1: Sending command: mosh-server 'new' '-c' '256' '-s' '-l' 'LANG=en_US.UTF-8' '-l' 'LANGUAGE=' '-l' 'LC_COLLATE=C' '--' 'command'
debug1: client_input_channel_req: channel 0 rtype exit-signal reply 0
debug1: channel 0: free: client-session, nchannels 1
Connection to 1.2.3.4 closed.
Transferred: sent 3840, received 1952 bytes, in 0.1 seconds
Bytes per second: sent 62372.0, received 31705.8
debug1: Exit status -1
/usr/bin/mosh: Did not find mosh server startup message. (Have you installed mosh on your server?)
```


Here is an abridged and commented excerpt of my `strace` output reflecting the kill (261374: parent process, 261377: child process):
```strace
/* master process: poll FDs 0 and 4: */
261374 18:16:27.744403 ppoll([{fd=0<socket:[1470497]>, events=POLLIN}, {fd=4</dev/pts/ptmx>, events=POLLIN}], 2, {tv_sec=60, tv_nsec=0}, NULL, 0 <unfinished ...>

/* child process only: utmp, passwd, chdir, execve(bash), execve(mosh-server) */
/* [numerous syscalls abridged] */

/* child process: mosh's regular syscalls */
261377 18:16:27.771760 close(3</etc/ssl/openssl.cnf>) = 0 <0.000024>
261377 18:16:27.771890 futex(0xffffa4152b70, FUTEX_WAKE_PRIVATE, 2147483647) = 0 <0.000018>
261377 18:16:27.771980 futex(0xffffa4152c44, FUTEX_WAKE_PRIVATE, 2147483647) = 0 <0.000018>
261377 18:16:27.772251 futex(0xffffa4152b68, FUTEX_WAKE_PRIVATE, 2147483647) = 0 <0.000019>
261377 18:16:27.772348 futex(0xffffa4152b6c, FUTEX_WAKE_PRIVATE, 2147483647) = 0 <0.000017>

/* master process: ppol() returns */
261374 18:16:27.773573 <... ppoll resumed>) = 1 ([{fd=0, revents=POLLIN}], left {tv_sec=59, tv_nsec=970879211}) <0.029146>
261374 18:16:27.773638 setitimer(ITIMER_REAL, {it_interval={tv_sec=0, tv_usec=0}, it_value={tv_sec=3600, tv_usec=0}}, {it_interval={tv_sec=0, tv_usec=0}, it_value={tv_sec=3599, tv_usec=965353}}) = 0 <0.000015>
261374 18:16:27.773701 read(0<socket:[1470497]>, "\f5'\213\20}V\20\341\37F\244\300\225\37\330A\202\333\360\1y'\301O\215p\356\33\204Qnj\265\316\300", 35000) = 36 <0.000024>

/* master process: close ptmx: */
261374 18:16:27.773785 close(4</dev/pts/ptmx>) = 0 <0.000055>
/* child process: get SIGHUP */
261377 18:16:27.773887 --- SIGHUP {si_signo=SIGHUP, si_code=SI_KERNEL} ---

261374 18:16:27.773911 ppoll([{fd=0<socket:[1470497]>, events=POLLIN}, {fd=4, events=POLLIN}], 2, {tv_sec=60, tv_nsec=0}, NULL, 0) = 1 ([{fd=4, revents=POLLNVAL}], left {tv_sec=59, tv_nsec=999995560}) <0.000037>
261374 18:16:27.774064 read(4, 0xaaaaabcd5cb1, 16375) = -1 EBADF (Bad file descriptor) <0.000025>
261374 18:16:27.774147 pipe2([4<pipe:[1469833]>, 5<pipe:[1469833]>], 0) = 0 <0.000046>
261374 18:16:27.774279 fcntl(4<pipe:[1469833]>, F_SETFD, FD_CLOEXEC) = 0 <0.000021>
261374 18:16:27.774366 fcntl(4<pipe:[1469833]>, F_GETFL) = 0 (flags O_RDONLY) <0.000019>
261374 18:16:27.774431 fcntl(4<pipe:[1469833]>, F_SETFL, O_RDONLY|O_NONBLOCK) = 0 <0.000019>
261374 18:16:27.774494 fcntl(5<pipe:[1469833]>, F_SETFD, FD_CLOEXEC) = 0 <0.000018>
261374 18:16:27.774556 fcntl(5<pipe:[1469833]>, F_GETFL) = 0x1 (flags O_WRONLY) <0.000028>
261374 18:16:27.774633 fcntl(5<pipe:[1469833]>, F_SETFL, O_WRONLY|O_NONBLOCK) = 0 <0.000055>
261374 18:16:27.774736 rt_sigaction(SIGCHLD, {sa_handler=0xaaaaabc22008, sa_mask=[CHLD], sa_flags=SA_RESTART}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0 <0.000024>
261374 18:16:27.774815 wait4(3342, 0xfffff7359f34, WNOHANG, NULL) = 0 <0.000023>
261374 18:16:27.774885 ppoll([{fd=0<socket:[1470497]>, events=POLLIN}, {fd=4<pipe:[1469833]>, events=POLLIN}], 2, {tv_sec=60, tv_nsec=0}, NULL, 0 <unfinished ...>

/* child process: SIGHUP is fatal */
261377 18:16:27.774949 +++ killed by SIGHUP +++

/* master process: cleanup after mosh-server got killed */
261374 18:16:27.774963 <... ppoll resumed>) = ? ERESTARTNOHAND (To be restarted if no handler) <0.000046>
261374 18:16:27.774994 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=3342, si_uid=1000, si_status=SIGHUP, si_utime=0, si_stime=0} ---
261374 18:16:27.775036 write(5<pipe:[1469833]>, "\0", 1) = 1 <0.000023>
```

**Comment by janmojzis** [comment]
Now it has the same problem, which can be triggered:
~~~
ssh -n -tt localhost date
~~~
as described at the beginning https://github.com/mobile-shell/mosh/issues/819#issue-186430918


This is what happens:

client requests terminal (-tt)
sends SSH_MSG_CHANNEL_REQUEST pty-req

client redirects stdin from /dev/null (-n)
sends SSH_MSG_CHANNEL_EOF

... and tinysshd closes the terminal filedescriptor, which sends a SIGHUP to the child process, and child process is killed.


Here is a workaround for testing (EOF is simply ignored in term. mode):
https://github.com/janmojzis/tinyssh/commit/198ee54e33fe0d1245a6cd556a8ea2a01979c498

But it will needs more serious 'SSH_MSG_CHANNEL_EOF' handling in terminal mode.

**Comment by xavierog** [comment]
> Here is a workaround for testing (EOF is simply ignored in term. mode): [janmojzis/tinyssh@198ee54](https://github.com/janmojzis/tinyssh/commit/198ee54e33fe0d1245a6cd556a8ea2a01979c498)

I applied this patch on top of the forkptyfix branch, compiled and tested multiple times: I did not get a single failure.

**Comment by janmojzis** [comment]
thanks for testing!
both changes are merged into the master
and they will be in the latest upcoming version 20240101

**Comment by xavierog** [comment]
Thanks for fixing!
The mosh workaround was acceptable, but mosh+tinysshd working out of the box is definitely better.
